BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m

```diff
@@ -83,12 +83,32 @@ + (nonnull NSString *)URLStringWithoutQueryForComponents:(nonnull NSURLComponent
     return string;
 }
 
+- (NSURLRequest *)findRequestFromTask:(NSURLSessionTask *)task andMetrics:(NSURLSessionTaskMetrics *)metrics
+API_AVAILABLE(macosx(10.12), ios(10.0), watchos(3.0), tvos(10.0)) {
+    NSURLRequest *req = task.originalRequest;
+    if (req != nil) {
+        return req;
+    }
+
+    NSArray<NSURLSessionTaskTransactionMetrics *> *transactionMetrics = metrics.transactionMetrics;
+    if (transactionMetrics.count > 0) {
+        // Try to find a phase with a request HTTP body.
+        for (NSUInteger i = 0; i < transactionMetrics.count; i++) {
+            req = transactionMetrics[i].request;
+            if (req.HTTPBody.length > 0) {
+                return req;
+            }
+        }
+        // Failing that, just use the last phase.
+        req = transactionMetrics[transactionMetrics.count - 1].request;
+    }
+
+    return req;
+}
+
```

Comment:
Why does this logic need to be so complicated?

If we have no way to reproduce a nil `originalRequest`, is it even worth having untested logic?

Reply:
I wanted to try harder to get a valid request. The `transactionMetrics` array can contain a whole bunch of different objects depending on what phase of the request it was in (including retries from a 3xx response). If you think it's too risky I can just go back to the old way.

Reply:
The NSURLSession header says:

```
@interface NSURLSessionTask : NSObject <NSCopying, NSProgressReporting>

@property (readonly)                 NSUInteger    taskIdentifier;    /* an identifier for this task, assigned by and unique to the owning session */
@property (nullable, readonly, copy) NSURLRequest  *originalRequest;  /* may be nil if this is a stream task */
@property (nullable, readonly, copy) NSURLRequest  *currentRequest;   /* may differ from originalRequest due to http server redirection */
@property (nullable, readonly, copy) NSURLResponse *response;         /* may be nil if no response has been received */
```

So we may be getting a nil `originalRequest` if it was a "stream task" - does that give a lead to find a way to reproduce the crash?

If so and this logic can't find a request for that case, probably better to keep it simple like:

```
        NSURLRequest *req = task.originalRequest ?: task.currentRequest;
        if (!req) {
            return;
        }
```

Reply:
OK, I've put the simplified version in. I couldn't repro the crash using a stream task unfortunately.